vscode/src/projectSystem.ts

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { log } from "qsharp-lang";
+import { Utils, URI } from "vscode-uri";
+import * as vscode from "vscode";
+
+// This flag is a hacky way of disabling the debugger when running in project mode.
+// We will have debugger support for projects _very soon_, and this flag will get
+// removed when that lands. To be clear, this is _not_ the ideal way to transmit
+// data to the debugger, but it is a temporary workaround.
+// It stores files that are a part of projects and disables the debugger on them.
+let PROJECT_MODE = false;
+export const getProjectMode = () => PROJECT_MODE;
+
+/**
+ * Finds and parses a manifest. Returns `null` if no manifest was found for the given uri, or if the manifest
+ * was malformed.
+ */
+export async function getManifest(uri: string): Promise<{
+  excludeFiles: string[];
+  excludeRegexes: string[];
+  manifestDirectory: string;
+} | null> {
+  const manifestDocument = await findManifestDocument(uri);
+
+  if (manifestDocument === null) {
+    PROJECT_MODE = false;
+    return null;
+  }
+
+  let parsedManifest;
+  try {
+    parsedManifest = JSON.parse(manifestDocument.content);
+  } catch (e) {
+    log.error(
+      "Found manifest document, but the Q# manifest was not valid JSON",
+      e,
+    );
+    PROJECT_MODE = false;
+    return null;
+  }
+
+  const manifestDirectory = Utils.dirname(manifestDocument.uri);
+
+  PROJECT_MODE = true;
+  return {
+    excludeFiles: parsedManifest.excludeFiles || [],
+    excludeRegexes: parsedManifest.excludeRegexes || [],
+    manifestDirectory: manifestDirectory.toString(),
+  };
+}
+
+/** Returns the manifest document if one is found
+ * returns null otherwise
+ */
+async function findManifestDocument(
+  currentDocumentUriString: string,
+): Promise<{ uri: vscode.Uri; content: string } | null> {
+  // /home/foo/bar/document.qs
+  const currentDocumentUri = URI.parse(currentDocumentUriString);
+
+  // /home/foo/bar
+  let uriToQuery = Utils.dirname(currentDocumentUri);
```

Comment:
Does this return JUST the directory portion? Do you lose the scheme and authority from the Uri in the process? Being the result of this is what you build the URIs to query later on, are you sure this works when the scheme and authority aren't important information? (Such as in a vscode.dev virtual file system, when opening a github repo by pressing '.', etc.)